services/httpd/handler_test.go

```diff
@@ -926,6 +927,93 @@ func TestHandler_XRequestId(t *testing.T) {
 	}
 }
 
+func TestThrottler_Handler(t *testing.T) {
+	t.Run("OK", func(t *testing.T) {
+		throttler := httpd.NewThrottler(2, 98)
+
+		// Send the total number of concurrent requests to the channel.
+		var concurrentN int32
+		concurrentCh := make(chan int)
+
+		h := throttler.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			atomic.AddInt32(&concurrentN, 1)
+			concurrentCh <- int(atomic.LoadInt32(&concurrentN))
+			time.Sleep(1 * time.Millisecond)
+			atomic.AddInt32(&concurrentN, -1)
+		}))
+
+		// Execute requests concurrently.
+		const n = 100
+		for i := 0; i < n; i++ {
+			go func() { h.ServeHTTP(nil, nil) }()
+		}
+
+		// Read the number of concurrent requests for every execution.
+		for i := 0; i < n; i++ {
+			if v := <-concurrentCh; v > 2 {
+				t.Fatalf("concurrent requests exceed maximum: %d", v)
+			}
+		}
+	})
+
+	t.Run("ErrTimeout", func(t *testing.T) {
+		throttler := httpd.NewThrottler(2, 1)
+		throttler.EnqueueTimeout = 1 * time.Millisecond
+
+		resp := make(chan struct{})
+		h := throttler.Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			resp <- struct{}{}
+		}))
+
+		// First two requests should execute immediately.
+		go func() { h.ServeHTTP(nil, nil) }()
+		go func() { h.ServeHTTP(nil, nil) }()
+		time.Sleep(100 * time.Millisecond)
```

Comment:
These sorts of sleeps are really racy on things like CircleCI. Maybe have a pair of starting/ending channels where the handler sends to starting and then sends to ending. That way, instead of a sleep here, you can read from starting twice to ensure that there are to goroutines inside of the handler.

Reply:
Updated to use a channel in 8a74c67.